main/src/io/id_tag_utils.hpp

```diff
@@ -0,0 +1,144 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*! @file
+ * @brief  CPU/GPU Particle ID tag utilities
+ *
+ * @author Christopher Bignamini <christopher.bignamini@gmail.com>
+ * @author Sebastian Keller <sebastian.f.keller@gmail.com>
+ */
+
+#pragma once
+
+#include <numeric>
+#include <span>
+#include <vector>
+
+#include "cstone/tree/definitions.h"
+#include "sph/types.hpp"
+
+namespace sphexa
+{
+
+using IdType = uint64_t;
+using LocalIndex = uint32_t;
+using CoordinateType = sph::SphTypes::CoordinateType;
+
+/*! @brief Number of bits used for tagging information storage
+ */
+constexpr IdType taggingMaskSize = 10;
+
+/*! @brief Given the taggingMaskSize, this is the maximum selection group id value plus one
+ */
+constexpr IdType supGroupId = (1 << taggingMaskSize) - 1;
+
+/*! @brief First tagging bit position
+ */
+constexpr IdType taggingMaskStartingBit = sizeof(IdType)*8 - taggingMaskSize;
+
+constexpr IdType taggingCheckMask = supGroupId << taggingMaskStartingBit;
+
+
+/*! @brief Tagged id identification condition functor
+ */
+struct IsMasked
+{
+    HOST_DEVICE_FUN uint8_t operator()(IdType id) const
+    {
+        return (id & taggingCheckMask) != 0;
+    }
+};
+
+/*! @brief Application of tagging mask to a given id
+ *
+ * @param[in]  groupId      selection group id (must be < supGroupId)
+ * @param[out] id           input id with tagging mask applied
+ */
+void applyTaggingMask(IdType groupId, IdType& id);
+
+/*! @brief Tagged id (in first:last range) identification, CPU version
+ *
+ * @param[in]  ids          ordered id list
+ * @param[in]  first        first id index // TODO number of elements and pass iterator?
+ * @param[in]  last         last (excluded) id index
+ * @param[out] taggedIdsIndexes  vector of indexes of tagged ids
+ */
+template<class IdTypeP, class LocalIndexP>
+void findTaggedIds(std::span<const IdTypeP> ids, size_t first, size_t last, std::vector<LocalIndexP>& taggedIdsIndexes);
+extern template void findTaggedIds<IdType, LocalIndex>(std::span<const IdType> ids, size_t first, size_t last, std::vector<LocalIndex>& taggedIdsIndexes);
+
+/*! @brief Tagged id (in first:last range) identification, GPU version
+ *
+ * @param[in]  ids          ordered id list
+ * @param[in]  first        first id index // TODO number of elements and pass iterator?
+ * @param[in]  last         last (excluded) id index
+ * @param[out] taggedIdsIndexes  vector of indexes of tagged ids
+ */
+template<class IdTypeP, class LocalIndexP>
+void findTaggedIdsGPU(std::span<const IdTypeP> ids, size_t first, size_t last, std::vector<LocalIndexP>& taggedIdsIndexes);
+extern template void findTaggedIdsGPU<IdType, LocalIndex>(std::span<const IdType> ids, size_t first, size_t last, std::vector<LocalIndex>& taggedIdsIndexes);
```

Comment:
Misunderstanding: the template instantiation belongs into .cu or .cpp file. Here you only need:
```c++
template<class IdTypeP, class LocalIndexP>
extern void findTaggedIdsGPU(std::span<const IdTypeP> ids, size_t first, size_t last, std::vector<LocalIndexP>& taggedIdsIndexes);
```